src/main/java/org/checkerframework/specimin/UnsolvedSymbolVisitor.java

```diff
@@ -0,0 +1,300 @@
+package org.checkerframework.specimin;
+
+import com.github.javaparser.ast.ImportDeclaration;
+import com.github.javaparser.ast.NodeList;
+import com.github.javaparser.ast.body.Parameter;
+import com.github.javaparser.ast.expr.MethodCallExpr;
+import com.github.javaparser.ast.expr.ObjectCreationExpr;
+import com.github.javaparser.ast.visitor.ModifierVisitor;
+import com.github.javaparser.ast.visitor.Visitable;
+import com.github.javaparser.resolution.UnsolvedSymbolException;
+import java.io.BufferedWriter;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * The visitor for the preliminary phase of Specimin. This visitor goes through the input files,
+ * spots out all the methods belonging to classes not in the source codes, and creates synthetic
+ * versions of those classes and methods. This preliminary step helps to prevent
+ * UnsolvedSymbolException errors for the next phases.
+ */
+public class UnsolvedSymbolVisitor extends ModifierVisitor<Void> {
+
+  /** List of classes not in the source codes */
+  private List<UnsolvedClass> missingClass;
+
+  /** The same as the root being used in SpeciminRunner */
+  private String rootDirectory;
+
+  /**
+   * This is to check if the current synthetic files are enough to prevent UnsolvedSymbolException
+   * or we still need more.
+   */
+  private boolean gotException;
+
+  /**
+   * The list of classes that have been created. We use this list to delete all the temporary
+   * synthetic classes when Specimin finishes its run
+   */
+  private Set<Path> createdClass;
+
+  /** List of import statement from the current compilation unit that is being visited */
+  private List<String> importStatement;
+
+  /**
+   * This map the classes in the compilation unit with the related import statements in that unit
+   */
+  private Map<String, String> classAndImportMap;
+
+  /**
+   * If there is any import statement that ends with *, this string will be replaced by one of the
+   * class from those import statements.
+   */
+  private String chosenPackage = "To.Be.Replaced";
```

Comment:
Is this a constant? If so, it should be `final` and its name should be `ALL_UPPER_CASE`.

Reply:
Professor, it is not a constant. It is supposed to be replaced with import statements ending with *. As we discussed in our meeting, if a user has a bunch of import statements ending with * and Specimin can't figure out the corresponding import statement for a class, then Specimin will simply pick one import statement ending with * and assume that the unknown class belongs to that import statement. I set its value as `"To.Be.Replaced"` because if I don't give it any value, the compiler will throw an uninitialized field error. But it will definitely be replaced by an import statement ending with *. If there is no import statements ending with *, then this variable just stay there and do nothing.

Reply:
Professor, for now, this string will be replaced every time we meet a * import statement during our iteration. So if there are more than  * import statement, this string will be equal to the last * import statement. Given the fact that Specimin will give a warning to users if they use * import statements, I think it doesn't matter which * import this `chosenPackage` be equal to. This is the piece of code where the string is replaced:

```
  private void setClassAndImportMap() {
    for (String importStatement : this.importStatement) {
      String[] importParts = importStatement.split("\\.");
      if (importParts.length > 0) {
        String className = importParts[importParts.length - 1];
        String packageName = importStatement.replace("." + className, "");
        if (className.equals("*")) {
          chosenPackage = packageName;
        } else {
          this.classAndImportMap.put(className, packageName);
        }
      }
    }
  }
```

Reply:
> So if there are more than one * import statements, this string will be equal to the last * import statement

This still seems problematic for me. I think it should be an error if this string is non-null when it is assigned (meaning there is a limit of one * import per class), or we should redesign this mechanism so that multiple * imports are supported. Silently getting the package name wrong for some * imports is a problem that will eventually cause us real problems.

Reply:
Professor,

I don't understand your concern. Maybe my documentation is misleading. The purpose of this whole thing is simply to know the package directories of imported classes. We do this by analyzing the import statements. For example, if we have these two import statements in the input file:

```
import org.github.sample.Name;
import org.hello.World;
```

Then our `classAndImportMap` will save two keys, which are two classes: `Name` and `World`. The value of each key is the package name in that import statement. So the value of key `Name` is `org.github.sample.Name`, and the value of key `World` is `org.hello.World`.

Thus, if we later realized that we need to create synthetic files for `Name` or `World`, then we can use the `classAndImportMap` to know the package name of that class in order to create the corresponding directory. For example, based on the map, we know that the directory of the synthetic file for class `Name` must be `org/github/sample/Name.java`.

And if our user uses some * import statement, then it is not possible for Specimin to figure out the directory of a missing class. In that case, what Specimin can do is to pick one * import statement randomly (in this case, the last one) and give user a warning.

Reply:
My concern is about this part:
> what Specimin can do is to pick one * import statement randomly (in this case, the last one) and give user a warning

I don't think Specimin should pick one randomly. I think Specimin should actively fail (i.e., throw an exception and stop) if there is ambiguity is which * import to choose. If we do not do this, I am concerned that Specimin will eventually produce synthetic classes that are in the wrong package (because the wrong * import was chosen for them). I don't want that to happen, because I think it will cause problems in some of the use cases I have in mind for Specimin.